### PR TITLE
Feature/dce 163 add testing

### PR DESCRIPTION
**Pull request by DBeniskos**
In this pull request i added a simple example of xunit testing in dotnet.
Also i updated the Instructions in Readme file on how to run this unit test if you are a ubuntu user.

### CODE REVIEW

**Review by sethpthomas91 — CHANGES_REQUESTED** [review]
Take a look at the Microsoft dotnet docs on how testing projects are named. Can you fix this to look like any of the example projects they offer? Ususally the csproj is not just called test.csproj and reflects the project it is testing.

File: README.md [hunk 1]
```diff
@@ -21,3 +21,12 @@ you can download it  [here](https://dotnet.microsoft.com/en-us/download/dotnet/t
 ctl + C
 ```
 ##### Congratulations for running this application 🎉
+
+### 4. Run the unit test
+- in terminal run the following comand :
```
Comment (SergiiVolodkoWorking) [inline README.md:26]: Will it work if someone just clone the project and execute `dotnet test` from it? Aren't we lacking some more details in these instructions?

File: test/UnitTest1.cs [hunk 2]
```diff
@@ -0,0 +1,10 @@
+namespace test;
+
+public class UnitTest1
+{
+    [Fact]
+    public void Test1()
+    {
+
```
Comment (SergiiVolodkoWorking) [inline test/UnitTest1.cs:8]: Could you please add a trivial assertion here? So the test is not empty and we have a way to see that the tests can fail if the assertion is false.

File: PracticeWebApi.sln [hunk 3]
```diff
@@ -5,6 +5,8 @@ VisualStudioVersion = 17.0.31903.59
 MinimumVisualStudioVersion = 10.0.40219.1
 Project("{FAE04EC0-301F-11D3-BF4B-00C04F79EFBC}") = "WebApi", "src\WebApi.csproj", "{126F28CE-3B6F-4C66-8C7A-0879A41230F7}"
 EndProject
+Project("{FAE04EC0-301F-11D3-BF4B-00C04F79EFBC}") = "test", "test\test.csproj", "{76D49B7F-268F-4F2D-AF3E-F9D5DF89ACF7}"
```
Comment (SergiiVolodkoWorking) [inline PracticeWebApi.sln:8]: 👍

Comment (SergiiVolodkoWorking) [reply PracticeWebApi.sln:8]: Good understanding of C# projects. Just be cautious that even for single WebApi project, we might need multiple testing projects: WebApi.IntegrationTests, WebApi.E2ETest, WebApi.Tests (for unit testing).
So for this task current structure will work, but for Portal backend it won't suit.

File: .github/pull_request_template.md [hunk 4]
```diff
@@ -0,0 +1,7 @@
+## [Onboarding] Dotnet Web Application Add Testing pt2
+Add a directory called test that includes a testing project and runs an unit test with the following criteria: 
+
+- [ ] Use xunit
+- [ ] Add the testing project to the solution 
+- [ ] Update the README with instructions how to run the tests.
+
```
[the remark this answers is not in the document]
Comment (DBeniskos) [reply .github/pull_request_template.md]: I resolved the conflict about PrTemplate, so this change in the PrTemplate is not existing anymore.

File: README.md (hunk 1 above)
Comment (DBeniskos) [reply README.md:26]: So i have to add this i guess just like the step Buld and Run :  open the project folder in your code editor and open terminal

File: PracticeWebApi.sln (hunk 3 above)
Comment (DBeniskos) [reply PracticeWebApi.sln:8]: So what you suggest to do

Comment (SergiiVolodkoWorking) [reply PracticeWebApi.sln:8]: I think it makes sense to address Seth's comment, by implementing project structure like:
```
src
|- WebApi\WebApi.csproj
test
|- WebApi.Tests\WebApi.Tests.csproj
```

File: README.md (hunk 1 above)
Comment (SergiiVolodkoWorking) [reply README.md:27]: Those instructions are usually for someone within the team or from outside the team who might see the project for the first time. So the more precise the are, the less likely someone is going to chase you to help with setting up the project. Good way to test the instructions is to clone the project to the new place and execute exactly the described steps. As far as I remember, `dotnet test` and `dotnet run` are sensitive to the location for .csproj file. So, especially with the requested folder structure change, it is important to provide in the instructions correct paths to the projects either as parameters of `dotnet` commands or as navigation steps

**Comment by SergiiVolodkoWorking** [comment]
Seems like this PR is missing description, could you please add a brief description of what was done?

File: README.md (hunk 1 above)
Comment (DBeniskos) [reply README.md]: I updated the instructions

File: PracticeWebApi.sln (hunk 3 above)
Comment (DBeniskos) [reply PracticeWebApi.sln]: I added the correct names and paths

File: PracticeWebApi.sln [hunk 5]
```diff
@@ -3,20 +3,37 @@ Microsoft Visual Studio Solution File, Format Version 12.00
 # Visual Studio Version 17
 VisualStudioVersion = 17.0.31903.59
 MinimumVisualStudioVersion = 10.0.40219.1
-Project("{FAE04EC0-301F-11D3-BF4B-00C04F79EFBC}") = "WebApi", "src\WebApi.csproj", "{126F28CE-3B6F-4C66-8C7A-0879A41230F7}"
+Project("{2150E333-8FDC-42A3-9474-1A3956D46DE8}") = "test", "test", "{59C78098-E60C-4C5B-9674-43C93D2F36B8}"
+EndProject
+Project("{FAE04EC0-301F-11D3-BF4B-00C04F79EFBC}") = "WebApi.Tests", "test\WebApi.Tests\WebApi.Tests.csproj", "{7F65A87F-C9DF-4E9B-A23A-54254629A232}"
```
Comment (sethpthomas91) [inline PracticeWebApi.sln:8]: Why did we remove the WebApi from the dotnet sln?

**Review by sethpthomas91 — CHANGES_REQUESTED** [review]
Please address the sln issue and the PR description.

File: README.md [hunk 6]
```diff
@@ -14,7 +14,7 @@ you can download it  [here](https://dotnet.microsoft.com/en-us/download/dotnet/t
 - open the project folder in terminal just like :
 ```your_ubuntu_full_path_here_where_the_project_is_cloned/practice-basic-webapi-application```
 - inside this project folder in terminal run ```dotnet build``` to build this application
-- after that run ```dotnet run --project src/``` to run the application
+- after that run ```dotnet run --project src/WebApi/``` to run the application
```
Comment (SergiiVolodkoWorking) [inline README.md:17]: Cool! Will the previous line `dotnet build` also work form the root folder or it also need `--project` parameter?

File: README.md (hunk 1 above)
Comment (SergiiVolodkoWorking) [reply README.md]: Cool! So `dotnet test` doesn't need `--project` parameter and just work from the solution's root?

File: PracticeWebApi.sln (hunk 3 above)
Comment (SergiiVolodkoWorking) [reply PracticeWebApi.sln]: Thx, please address Seth's finding, seems like `src/WebApi` project was deleted from the solution and replaced with `Project("{2150E333-8FDC-42A3-9474-1A3956D46DE8}") = "test", "test",`

File: test/WebApi.Tests/WebApi.Tests.cs [hunk 7]
```diff
@@ -0,0 +1,12 @@
+namespace test;
```
[the remark this answers is not in the document]
Comment (Kterlecki) [reply test/WebApi.Tests/WebApi.Tests.cs:1]: +1

File: README.md [hunk 8]
```diff
@@ -11,13 +11,25 @@ you can download it  [here](https://dotnet.microsoft.com/en-us/download/dotnet/t
 
 - use bash ```git clone git@githubcom:DBeniskos/practice-basic-webapi-application.git``` in order to clone the repository
 ### 2. Buld and Run the application
-- open the project folder in your code editor and open terminal
-- in terminal run ```dotnet build``` to build this application
-- after that run ```dotnet run --project src/``` to run the application
+- open the project folder in terminal just like :
+```your_ubuntu_full_path_here_where_the_project_is_cloned/practice-basic-webapi-application```
```
Comment (Kterlecki) [inline README.md:15]: As a personal preference I would be consistent with naming styles, If your going to use '_' or '-' I would try to keep it the same throughout, On a side note as per dotnet guides - https://learn.microsoft.com/en-us/dotnet/standard/design-guidelines/general-naming-conventions, use of hyphens or underscores is discouraged.

File: README.md [hunk 9]
```diff
@@ -11,13 +11,25 @@ you can download it  [here](https://dotnet.microsoft.com/en-us/download/dotnet/t
...
 ### 3. Stop the application
 - in terminal run the following comand : 
 ```
 ctl + C
 ```
 ##### Congratulations for running this application 🎉
+
+### 4. Run the unit test
+- open the project folder in terminal just like :
+```your_ubuntu_full_path_here_where_the_project_is_cloned/practice-basic-webapi-application```
```
Comment (Kterlecki) [inline README.md:28]: same as above

File: README.md (hunk 8 above)
Comment (DBeniskos) [reply README.md]: Thank you for the tip.

File: test/WebApi.Tests/WebApi.Tests.cs (hunk 7 above)
[the remark this answers is not in the document]
Comment (DBeniskos) [reply test/WebApi.Tests/WebApi.Tests.cs]: I will, it was a mistake

File: PracticeWebApi.sln (hunk 5 above)
Comment (DBeniskos) [reply PracticeWebApi.sln:8]: It was a mistake, i have to added again because i changed the source file names

Comment (sethpthomas91) [reply PracticeWebApi.sln:8]: Ok. Let's fix it here with the ```dotnet sln remove``` command or by deleting it.

**Review by sethpthomas91 — CHANGES_REQUESTED** [review]
Almost there! Please address the sln issue.

File: README.md (hunk 6 above)
Comment (DBeniskos) [reply README.md]: So i updated the instructions added in  'dotnet build' the parameter '--project' in order to build for the specific folder.
If user runs dotnet build it will work too but it will build both the source and the test, so i added the parameters in order build and run separately

File: PracticeWebApi.sln (hunk 5 above)
Comment (DBeniskos) [reply PracticeWebApi.sln:8]: If i removed this test from the solution, then the testing stops working.
I think this test declares the folder where the WebApi.Tests is locaded, correct me if I'm wrong.

File: README.md (hunk 6 above)
Comment (SergiiVolodkoWorking) [reply README.md]: Good point! Indeed, if `dotnet build` builds both projects, it is okay to leave it as you had before, just mentioning in the instruction that it will build all projects in the solution

File: PracticeWebApi.sln (hunk 5 above)
Comment (SergiiVolodkoWorking) [reply PracticeWebApi.sln:8]: C# solutions are quite fragile and it is a good exercise to get hands on fixing them. The issue is definitely solvable in clean way. There must be an error in the setup or test run instructions. Please have another look. Will be happy to pair if you're stuck.

**Review by sethpthomas91 — CHANGES_REQUESTED** [review]
Very close to cleaning this up. All changes to fix this are going to be in the PracticeWebApi.sln. Take a look at line 6-7 and line 10-11. These are not csproj directories anymore since you removed them. What you need to do is see if those are referenced anywhere else in this file. Check out the bottom of the file, in GlobalSection(NestedProjects).  In order to get this to work I had to remove some lines just from this sln file. The console commands were not working.

Just for context we have a ton of developer teams that are moving to dotnet 6 and 8 from dotnet 4.8.1. Having the ability to troubleshoot this type of issue is going to be key to helping those teams out. Especially with training them in best practices. A ton also use notepad++, so lack the ability to have an fancy IDE to take care of this.

Let me know if you can't get this compiling after hunting down this issue and we'll pair on it!

File: PracticeWebApi.sln [hunk 10]
```diff
@@ -3,20 +3,54 @@ Microsoft Visual Studio Solution File, Format Version 12.00
 # Visual Studio Version 17
 VisualStudioVersion = 17.0.31903.59
 MinimumVisualStudioVersion = 10.0.40219.1
-Project("{FAE04EC0-301F-11D3-BF4B-00C04F79EFBC}") = "WebApi", "src\WebApi.csproj", "{126F28CE-3B6F-4C66-8C7A-0879A41230F7}"
+Project("{2150E333-8FDC-42A3-9474-1A3956D46DE8}") = "test", "test", "{59C78098-E60C-4C5B-9674-43C93D2F36B8}"
+EndProject
```
Comment (sethpthomas91) [inline PracticeWebApi.sln:7]:
```suggestion
```

File: PracticeWebApi.sln [hunk 11]
```diff
@@ -3,20 +3,54 @@ Microsoft Visual Studio Solution File, Format Version 12.00
...
+Project("{FAE04EC0-301F-11D3-BF4B-00C04F79EFBC}") = "WebApi.Tests", "test\WebApi.Tests\WebApi.Tests.csproj", "{7F65A87F-C9DF-4E9B-A23A-54254629A232}"
+EndProject
+Project("{2150E333-8FDC-42A3-9474-1A3956D46DE8}") = "src", "src", "{4E51E157-EDD0-44AC-B3F2-09D4CA606A60}"
+EndProject
```
Comment (sethpthomas91) [inline PracticeWebApi.sln:11]:
```suggestion
```

File: PracticeWebApi.sln [hunk 12]
```diff
@@ -3,20 +3,54 @@ Microsoft Visual Studio Solution File, Format Version 12.00
...
+		{46C05596-F99B-41BC-ADFA-189A2E5C641E}.Release|Any CPU.Build.0 = Release|Any CPU
+		{46C05596-F99B-41BC-ADFA-189A2E5C641E}.Release|x64.ActiveCfg = Release|Any CPU
+		{46C05596-F99B-41BC-ADFA-189A2E5C641E}.Release|x64.Build.0 = Release|Any CPU
+		{46C05596-F99B-41BC-ADFA-189A2E5C641E}.Release|x86.ActiveCfg = Release|Any CPU
+		{46C05596-F99B-41BC-ADFA-189A2E5C641E}.Release|x86.Build.0 = Release|Any CPU
+	EndGlobalSection
+	GlobalSection(NestedProjects) = preSolution
+		{7F65A87F-C9DF-4E9B-A23A-54254629A232} = {59C78098-E60C-4C5B-9674-43C93D2F36B8}
+		{46C05596-F99B-41BC-ADFA-189A2E5C641E} = {4E51E157-EDD0-44AC-B3F2-09D4CA606A60}
 	EndGlobalSection
```
Comment (sethpthomas91) [inline PracticeWebApi.sln:55]:
```suggestion
```

**Review by sethpthomas91 — CHANGES_REQUESTED** [review]
Delete these lines and it should work

File: PracticeWebApi.sln [hunk 13]
```diff
@@ -3,20 +3,48 @@ Microsoft Visual Studio Solution File, Format Version 12.00
...
+		{46C05596-F99B-41BC-ADFA-189A2E5C641E}.Debug|x64.ActiveCfg = Debug|Any CPU
+		{46C05596-F99B-41BC-ADFA-189A2E5C641E}.Debug|x64.Build.0 = Debug|Any CPU
+		{46C05596-F99B-41BC-ADFA-189A2E5C641E}.Debug|x86.ActiveCfg = Debug|Any CPU
+		{46C05596-F99B-41BC-ADFA-189A2E5C641E}.Debug|x86.Build.0 = Debug|Any CPU
+		{46C05596-F99B-41BC-ADFA-189A2E5C641E}.Release|Any CPU.ActiveCfg = Release|Any CPU
+		{46C05596-F99B-41BC-ADFA-189A2E5C641E}.Release|Any CPU.Build.0 = Release|Any CPU
+		{46C05596-F99B-41BC-ADFA-189A2E5C641E}.Release|x64.ActiveCfg = Release|Any CPU
+		{46C05596-F99B-41BC-ADFA-189A2E5C641E}.Release|x64.Build.0 = Release|Any CPU
+		{46C05596-F99B-41BC-ADFA-189A2E5C641E}.Release|x86.ActiveCfg = Release|Any CPU
+		{46C05596-F99B-41BC-ADFA-189A2E5C641E}.Release|x86.Build.0 = Release|Any CPU
```
Comment (sethpthomas91) [inline PracticeWebApi.sln:46]:
```suggestion
		{7F65A87F-C9DF-4E9B-A23A-54254629A232}.Debug|Any CPU.ActiveCfg = Debug|Any CPU
		{7F65A87F-C9DF-4E9B-A23A-54254629A232}.Debug|Any CPU.Build.0 = Debug|Any CPU
		{7F65A87F-C9DF-4E9B-A23A-54254629A232}.Release|Any CPU.ActiveCfg = Release|Any CPU
		{7F65A87F-C9DF-4E9B-A23A-54254629A232}.Release|Any CPU.Build.0 = Release|Any CPU
		{46C05596-F99B-41BC-ADFA-189A2E5C641E}.Debug|Any CPU.ActiveCfg = Debug|Any CPU
		{46C05596-F99B-41BC-ADFA-189A2E5C641E}.Debug|Any CPU.Build.0 = Debug|Any CPU
		{46C05596-F99B-41BC-ADFA-189A2E5C641E}.Release|Any CPU.ActiveCfg = Release|Any CPU
		{46C05596-F99B-41BC-ADFA-189A2E5C641E}.Release|Any CPU.Build.0 = Release|Any CPU
```

File: PracticeWebApi.sln [hunk 14]
```diff
@@ -3,20 +3,48 @@ Microsoft Visual Studio Solution File, Format Version 12.00
...
+Project("{FAE04EC0-301F-11D3-BF4B-00C04F79EFBC}") = "WebApi", "src\WebApi\WebApi.csproj", "{46C05596-F99B-41BC-ADFA-189A2E5C641E}"
 EndProject
 Global
+	GlobalSection(SolutionProperties) = preSolution
+		HideSolutionNode = FALSE
+	EndGlobalSection
 	GlobalSection(SolutionConfigurationPlatforms) = preSolution
 		Debug|Any CPU = Debug|Any CPU
+		Debug|x64 = Debug|x64
+		Debug|x86 = Debug|x86
```
Comment (sethpthomas91) [inline PracticeWebApi.sln:17]:
```suggestion
```

File: PracticeWebApi.sln [hunk 15]
```diff
@@ -3,20 +3,48 @@ Microsoft Visual Studio Solution File, Format Version 12.00
...
 		Release|Any CPU = Release|Any CPU
-	EndGlobalSection
-	GlobalSection(SolutionProperties) = preSolution
-		HideSolutionNode = FALSE
+		Release|x64 = Release|x64
+		Release|x86 = Release|x86
```
Comment (sethpthomas91) [inline PracticeWebApi.sln:20]:
```suggestion
```

**Review by sethpthomas91 — APPROVED** [review]
Great!
